File: android/src/ti/mailcore/TiMailcoreModule.java
/**
 * TiMailcore
 */

package ti.mailcore;

import org.appcelerator.titanium.TiApplication;

import org.appcelerator.kroll.annotations.Kroll;
import org.appcelerator.kroll.KrollDict;
import org.appcelerator.kroll.KrollFunction;
import org.appcelerator.kroll.KrollModule;
import org.appcelerator.kroll.common.TiConfig;
import org.appcelerator.kroll.common.Log;

import com.libmailcore.MailException;
import com.libmailcore.OperationCallback;
import com.libmailcore.ConnectionType;
import com.libmailcore.IMAPOperation;
import com.libmailcore.IMAPSession;
import com.libmailcore.IMAPMessage;
import com.libmailcore.IMAPFetchContentOperation;
import com.libmailcore.IMAPFetchMessagesOperation;
import com.libmailcore.IMAPFetchFoldersOperation;
import com.libmailcore.IMAPFolderInfoOperation;
import com.libmailcore.IMAPMessagesRequestKind;
import com.libmailcore.IMAPFolder;
import com.libmailcore.SMTPSession;
import com.libmailcore.SMTPOperation;
import com.libmailcore.OperationCallback;
import com.libmailcore.IndexSet;
import com.libmailcore.Range;
import com.libmailcore.Address;
import com.libmailcore.AbstractMessage;
import com.libmailcore.MessageHeader;
import com.libmailcore.MessageParser;
import com.libmailcore.MessageBuilder;
import com.libmailcore.Attachment;
import com.libmailcore.AbstractPart;

import java.util.HashMap;
import java.util.ArrayList;

import android.util.Base64;

@Kroll.module(name="TiMailcore", id="ti.mailcore")
public class TiMailcoreModule extends KrollModule {
	private static final String LCAT = "TiMailcoreModule";
	private static final boolean DBG = TiConfig.LOGD;

	public TiMailcoreModule() {
		super();
	}

	@Kroll.onAppCreate
	public static void onAppCreate(TiApplication app) {}

	private Object _applyCredentials(KrollDict credentials, boolean smpt) {
		String email = (String)credentials.get("email");
		String password = (String)credentials.get("password");
		String host_name = (String)credentials.get("host");
		int port_name = credentials.getInt("port");
		int ctype = credentials.containsKey("ctype") ? credentials.getInt("ctype") : ConnectionType.ConnectionTypeTLS;
		String oauth_token = (credentials.get("oauth_token") != null) ? (String)credentials.get("oauth_token") : null;

		Object session;

		if(smpt) {
			SMTPSession s_session = new SMTPSession();
			s_session.setUsername(email);
			s_session.setPassword(password);
			s_session.setHostname(host_name);
			s_session.setPort(port_name);
			s_session.setConnectionType(ctype);
			if(oauth_token != null) {
				s_session.setOAuth2Token(oauth_token);
			}
			session = s_session;
		} else {
			IMAPSession i_session = new IMAPSession();
			i_session.setUsername(email);
			i_session.setPassword(password);
			i_session.setHostname(host_name);
			i_session.setPort(port_name);
			i_session.setConnectionType(ctype);
			if(oauth_token != null) {
				i_session.setOAuth2Token(oauth_token);
			}
			session = i_session;
		}
		return session;
	}

	private KrollDict _messageToJSON(AbstractMessage msg) {
		KrollDict email = compose();
		KrollDict email_headers = email.getKrollDict("headers");
		KrollDict email_addresses = email.getKrollDict("addresses");
		MessageHeader header = msg.header();

		if(header != null) {
			// Basic data and headers
			if(header.subject() != null) {
				email.put("subject", header.subject());
			}
			if(header.date() != null) {
				email.put("date", header.date().toString());
			}
			if(header.receivedDate() != null) {
				email.put("receivedDate", header.receivedDate().toString());
			}
			if(header.allExtraHeadersNames() != null) {
				for(String extra : header.allExtraHeadersNames()) {
					String extra_data = header.extraHeaderValueForName(extra);
					if(extra_data != null) {
						email_headers.put(extra, extra_data);
					}
				}
			}

			// 'From' address
			if(header.from() != null) {
				KrollDict from = email_addresses.getKrollDict("from");
				String display_name = header.from().displayName();
				String mailbox = header.from().mailbox();
				if(display_name != null) {
					from.put("name", display_name);
				}
				if(mailbox != null) {
					from.put("mailbox", mailbox);
				}
			}

			// Remainder of the address types
			KrollDict address_sections = new KrollDict();
			address_sections.put("to", header.to());
			address_sections.put("cc", header.cc());
			address_sections.put("bcc", header.bcc());
			address_sections.put("replyTo", header.replyTo());
			for(Object address_section_obj : address_sections.keySet()) {
				String address_section = (String)address_section_obj;
				java.util.List<Address> addresses = (java.util.List<Address>)(address_sections.get(address_section));

				if(addresses != null) {
					ArrayList my_addresses = new ArrayList();

					for(Address address : addresses) {
						KrollDict new_address = new KrollDict();

						String display_name = address.displayName();
						String mailbox = address.mailbox();
						if(display_name != null) {
							new_address.put("name", display_name);
						}
						if(mailbox != null) {
							new_address.put("mailbox", mailbox);
						}
						my_addresses.add(new_address);
					}

					email_addresses.put(address_section, my_addresses.toArray(new Object[my_addresses.size()]));
				}
			}
		}

		return email;
	}

	// Private methods defining email json structure
	private KrollDict _getEmailStructure() {
		KrollDict structure = new KrollDict();
		structure.put("subject", "");
		structure.put("body", "");
		return structure;
	}
	private KrollDict _getHeaderStructure() {
		KrollDict structure = new KrollDict();
		return structure;
	}
	private KrollDict _getAddressStructure() {
		KrollDict structure = new KrollDict();
		structure.put("to", new Object[0]);
		structure.put("cc", new Object[0]);
		structure.put("bcc", new Object[0]);
		structure.put("from", new KrollDict());
		structure.put("replyTo", new Object[0]);
		return structure;
	}

	private void _applyHeader(KrollDict header, KrollDict email) {
		email.put("headers", header);
	}

	private void _applyAddress(KrollDict address, KrollDict email) {
		email.put("addresses", address);
	}

	// Abstract base class of all callback job handlers. Just have to handle
	// creation of the operation and formatting of its results.
	private abstract class CallbackCaller<O, S> implements OperationCallback {
		private O operation;
		protected KrollFunction callback;
		private S session;

		public CallbackCaller(KrollDict credentials, KrollFunction cb, boolean smpt) {
			callback = cb;
			session = (S)(_applyCredentials(credentials, smpt));
		}

		abstract protected void operationStart(O op); // Because java is stupid and cant call methods on generics
		abstract protected O createOperation(S session);
		abstract protected Object formatResult(O operation);

		public void start() {
			operation = createOperation(session);
			operationStart(operation);
		}

		public void succeeded() {
			callback.call(getKrollObject(), new Object[]{null, formatResult(operation)});
		}

		public void failed(MailException exception) {
			callback.call(getKrollObject(), new Object[]{exception.getMessage(), null});
		}
	}

	// IMAP and SMTP specific versions
	private abstract class CallbackCallerIMAP extends CallbackCaller<IMAPOperation, IMAPSession> {
			public CallbackCallerIMAP(KrollDict credentials, KrollFunction cb, boolean smpt) {
				super(credentials, cb, false);
			}

			protected void operationStart(IMAPOperation op) {
				op.start(this);
			}
	}
	private abstract class CallbackCallerSMTP extends CallbackCaller<SMTPOperation, SMTPSession> {
			public CallbackCallerSMTP(KrollDict credentials, KrollFunction cb, boolean smpt) {
				super(credentials, cb, true);
			}

			protected void operationStart(SMTPOperation op) {
				op.start(this);
			}
	}

	// Get list of folders
	@Kroll.method
	public void getFolders(KrollDict imap, KrollFunction cb) {
		CallbackCaller caller = new GetFoldersCaller(imap, cb);
		caller.start();
	}
	private class GetFoldersCaller extends CallbackCallerIMAP {
		public GetFoldersCaller(KrollDict credentials, KrollFunction cb) {
			super(credentials, cb, false);
		}

		protected IMAPOperation createOperation(IMAPSession session) {
			return session.fetchAllFoldersOperation();
		}

		protected Object formatResult(IMAPOperation operation) {
			java.util.List<IMAPFolder> folders = ((IMAPFetchFoldersOperation)operation).folders();
			ArrayList result = new ArrayList();
			for(IMAPFolder folder : folders) {
				result.add(folder.path());
			}
			return result.toArray(new String[result.size()]);
		}
	}


	// Get Info for a specific folder
	@Kroll.method
	public void getFolderInfo(KrollDict imap, String folder, KrollFunction cb) {
		CallbackCaller caller = new GetFolderInfoCaller(imap, folder, cb);
		caller.start();
	}
	private class GetFolderInfoCaller extends CallbackCallerIMAP {
		private String folder;

		public GetFolderInfoCaller(KrollDict credentials, String f, KrollFunction cb) {
			super(credentials, cb, false);
			folder = f;
		}

		protected IMAPOperation createOperation(IMAPSession session) {
			return session.folderInfoOperation(folder);
		}

		protected Object formatResult(IMAPOperation operation) {
			IMAPFolderInfoOperation info_op = (IMAPFolderInfoOperation)operation;

			KrollDict result = new KrollDict();
			result.put("UIDNEXT", info_op.info().uidNext());
			result.put("UIDVALIDITY", info_op.info().uidValidity());
			result.put("HIGHESTMODSEQ", info_op.info().modSequenceValue());
			result.put("messages_count", info_op.info().messageCount());

			return result;
		}
	}

	// Get basic info of a mail folder with optional uid range
	@Kroll.method
	public void getMail(KrollDict imap, String folder, long range[], KrollFunction cb) {
		if(range == null) {
			range = new long[]{1, Long.MAX_VALUE};
		}
		IndexSet uids = IndexSet.indexSetWithRange(new Range(range[0], range[1] - range[0]));
		CallbackCaller caller = new GetMailCaller(imap, cb, folder, uids);
		caller.start();
	}
	private class GetMailCaller extends CallbackCallerIMAP {
		private String folder;
		private IndexSet uids;

		public GetMailCaller(KrollDict credentials, KrollFunction cb, String f, IndexSet u) {
			super(credentials, cb, false);
			folder = f;
			uids = u;
		}

		protected IMAPOperation createOperation(IMAPSession session) {
			int requestKind = IMAPMessagesRequestKind.IMAPMessagesRequestKindHeaders |
			IMAPMessagesRequestKind.IMAPMessagesRequestKindHeaderSubject |
			IMAPMessagesRequestKind.IMAPMessagesRequestKindExtraHeaders |
			IMAPMessagesRequestKind.IMAPMessagesRequestKindStructure;

			IMAPFetchMessagesOperation op = session.fetchMessagesByUIDOperation(folder, requestKind, uids);

			// TODO - should be passed in
			ArrayList extra_headers = new ArrayList();
			extra_headers.add("Received-SPF");
			op.setExtraHeaders(extra_headers);
			return op;
		}

		protected Object formatResult(IMAPOperation operation) {
			java.util.List<IMAPMessage> messages = ((IMAPFetchMessagesOperation)operation).messages();
			ArrayList result = new ArrayList();
			for(IMAPMessage message : messages) {
				KrollDict email_result = new KrollDict();
				email_result.put("uid", message.uid());
				email_result.put("sender_name", (message.header().sender().displayName() != null) ? message.header().sender().displayName() : "");
				email_result.put("sender_mailbox", (message.header().sender().mailbox() != null) ? message.header().sender().mailbox() : "");
				email_result.put("subject", (message.header().subject() != null) ? message.header().subject() : "");
				email_result.put("received_time", (message.header().receivedDate() != null) ? message.header().receivedDate().toString() : "");
				email_result.put("has_attachments", (message.attachments() != null) && message.attachments().size() > 0 ? true : false);

				for(String hname : message.header().allExtraHeadersNames()) {
					String extra_header = message.header().extraHeaderValueForName(hname);
					if(extra_header != null) {
						email_result.put(hname, extra_header);
					}
				}

				result.add(email_result);
			}

			return result.toArray(new KrollDict[result.size()]);
		}
	}

	// Get detailed info about one specific email by its uid
	@Kroll.method
	public void getMailInfo(KrollDict imap, String folder, long uid, KrollFunction cb) {
		CallbackCaller caller = new GetMailInfoCaller(imap, cb, uid, folder);
		caller.start();
	}
	private class GetMailInfoCaller extends CallbackCallerIMAP {
		private String folder;
		private long uid;

		public GetMailInfoCaller(KrollDict credentials, KrollFunction cb, long u, String f) {
			super(credentials, cb, false);
			uid = u;
			folder = f;
		}

		protected IMAPOperation createOperation(IMAPSession session) {
			return session.fetchMessageByUIDOperation(folder, uid);
		}

		protected Object formatResult(IMAPOperation operation) {
			byte[] data = ((IMAPFetchContentOperation)operation).data();
			MessageParser parser = MessageParser.messageParserWithData(data);

			KrollDict email = _messageToJSON(parser);
			if(parser != null) {
				email.put("body", parser.htmlBodyRendering());
				if(parser.attachments() != null) {
					ArrayList attachments = new ArrayList();
					for(AbstractPart attachment_abstract : parser.attachments()) {
						Attachment attachment = (Attachment)attachment_abstract;
						KrollDict attachment_object = new KrollDict();

						attachment_object.put("file_name", attachment.filename() == null ? "" : attachment.filename());
						attachment_object.put("mime_type", attachment.mimeType() == null ? "" : attachment.mimeType());
						attachment_object.put("data", attachment.data() == null ? "" : Base64.encodeToString(attachment.data(), Base64.NO_WRAP));

						attachments.add(attachment_object);
					}
					email.put("attachments", attachments.toArray(new Object[attachments.size()]));
				}
			}
			return email;
		}
	}

	// Get a basic skeleton of an email
	@Kroll.method
	public KrollDict compose() {
		KrollDict email_data = _getEmailStructure();
		KrollDict email_header = _getHeaderStructure();
		KrollDict email_address = _getAddressStructure();

		_applyHeader(email_header, email_data);
		_applyAddress(email_address, email_data);

		return email_data;
	}


	// Send an email
	@Kroll.method
	public void send(KrollDict smtp, KrollDict email, KrollFunction cb) {
			KrollDict email_addresses = email.getKrollDict("addresses");
			String subject = (String)(email.get("subject"));
			KrollDict from = email_addresses.getKrollDict("from");

			int receiver_count = 0;
			if(email_addresses.get("to") != null) {
				receiver_count += ((Object[])(email_addresses.get("to"))).length;
			}
			if(email_addresses.get("cc") != null) {
				receiver_count += ((Object[])(email_addresses.get("cc"))).length;
			}
			if(email_addresses.get("bcc") != null) {
				receiver_count += ((Object[])(email_addresses.get("bcc"))).length;
			}

	    if((subject != null) && (from != null) && (receiver_count > 0)) {
				CallbackCaller caller = new SendEmailCaller(smtp, cb, email);
				caller.start();
			} else {
				Log.i("Sending email", "Cannot send email without a subject, a 'from' section, and some destination");
			}

	}
	private class SendEmailCaller extends CallbackCallerSMTP {
		private KrollDict email;
		private KrollDict sent_email;

		public SendEmailCaller(KrollDict credentials, KrollFunction cb, KrollDict e) {
			super(credentials, cb, true);
			email = e;
			sent_email = new KrollDict();
		}

		protected SMTPOperation createOperation(SMTPSession session) {
			KrollDict email_headers = email.getKrollDict("headers");
			KrollDict email_addresses = email.getKrollDict("addresses");

			MessageBuilder builder = new MessageBuilder();

			String subject = (String)(email.get("subject"));
			KrollDict from = email_addresses.getKrollDict("from");

			// subject
			builder.header().setSubject(subject);

			// From
			builder.header().setFrom(Address.addressWithDisplayName((String)from.get("name"), (String)from.get("mailbox")));

			// To
			if(email_addresses.get("to") != null) {
				ArrayList to = new ArrayList();
				Object[] targets = ((Object[])(email_addresses.get("to")));
				if(targets != null) {
					for(Object t : targets) {
						HashMap target = (HashMap)t;
						to.add(Address.addressWithDisplayName((String)target.get("name"), (String)target.get("mailbox")));
					}
					builder.header().setTo(to);
				}
			}
			if(email_addresses.get("cc") != null) {
				ArrayList cc = new ArrayList();
				Object[] targets = ((Object[])(email_addresses.get("cc")));
				if(targets != null) {
					for(Object t : targets) {
						HashMap target = (HashMap)t;
						cc.add(Address.addressWithDisplayName((String)target.get("name"), (String)target.get("mailbox")));
					}
					builder.header().setCc(cc);
				}
			}
			if(email_addresses.get("bcc") != null) {
				ArrayList bcc = new ArrayList();
				Object[] targets = ((Object[])(email_addresses.get("bcc")));
				if(targets != null) {
					for(Object t : targets) {
						HashMap target = (HashMap)t;
						bcc.add(Address.addressWithDisplayName((String)target.get("name"), (String)target.get("mailbox")));
					}
					builder.header().setBcc(bcc);
				}
			}

			// Copy headers
			for(String key : email_headers.keySet()) {
				builder.header().setExtraHeader(key, (String)email_headers.get(key));
			}

			// Email body
			String body = (String)email.get("body");
			if(body != null) {
				builder.setHTMLBody(body);
			}

			sent_email = _messageToJSON(builder);
			return session.sendMessageOperation(builder.data());
		}

		protected Object formatResult(SMTPOperation operation) {
			return sent_email;
		}
	}
}

File: iphone/Classes/TiMailcoreModule.m
/**
 * TiMailcore
 */

#import "TiMailcoreModule.h"
#import "TiBase.h"
#import "TiHost.h"
#import "TiUtils.h"
#import "MailCore/MailCore.h"

@implementation TiMailcoreModule

#pragma mark Internal

// this is generated for your module, please do not change it
-(id)moduleGUID
{
	return @"12eadb4c-788a-4eac-b2f5-0a1ead4ca2c7";
}

// this is generated for your module, please do not change it
-(NSString*)moduleId
{
	return @"ti.mailcore";
}

#pragma mark Lifecycle

-(void)startup
{
	// this method is called when the module is first loaded
	// you *must* call the superclass
	[super startup];

	NSLog(@"[INFO] %@ loaded",self);
}

-(void)shutdown:(id)sender
{
	// this method is called when the module is being unloaded
	// typically this is during shutdown. make sure you don't do too
	// much processing here or the app will be quit forceably

	// you *must* call the superclass
	[super shutdown:sender];
}

#pragma mark Cleanup

-(void)dealloc
{
	// release any resources that have been retained by the module
	[super dealloc];
}

#pragma mark Internal Memory Management

-(void)didReceiveMemoryWarning:(NSNotification*)notification
{
	// optionally release any resources that can be dynamically
	// reloaded once memory is available - such as caches
	[super didReceiveMemoryWarning:notification];
}

#pragma mark Listener Notifications

-(void)_listenerAdded:(NSString *)type count:(int)count
{
	if (count == 1 && [type isEqualToString:@"my_event"])
	{
		// the first (of potentially many) listener is being added
		// for event named 'my_event'
	}
}

-(void)_listenerRemoved:(NSString *)type count:(int)count
{
	if (count == 0 && [type isEqualToString:@"my_event"])
	{
		// the last listener called for event named 'my_event' has
		// been removed, we can optionally clean up any resources
		// since no body is listening at this point for that event
	}
}

- (id)_applyCredentials:(NSDictionary *)credentials smpt:(BOOL)smpt {
    NSString * email = [credentials objectForKey:@"email"];
    NSString * password = [credentials objectForKey:@"password"];
    NSString * host_name = [credentials objectForKey:@"host"];
    NSString * oauth_token = [credentials objectForKey:@"oauth_token"];
    
    int port_name = [TiUtils intValue:[credentials objectForKey:@"port"]];
    MCOConnectionType ctype = [credentials objectForKey:@"ctype"] ? [TiUtils intValue:[credentials objectForKey:@"ctype"]] : MCOConnectionTypeTLS;
    
    id session;
    if(smpt) {
        session = [[MCOSMTPSession alloc] init];
    } else {
        session = [[MCOIMAPSession alloc] init];
    }
    
    [session setUsername: email];
    [session setPassword: password];
    [session setHostname: host_name];
    [session setPort: port_name];
    [session setConnectionType: ctype];
    if(oauth_token) {
        [session setOAuth2Token: oauth_token];
    }
    
    if(smpt == NO) {
        MCOIMAPOperation * op = [session checkAccountOperation];
        [op start:^(NSError * error) {
            if(error) {
                NSLog(@"[INFO] Connection appears to be invalid.");
            }
        }];
    }
    return session;
}

- (NSMutableDictionary*)_messageToJSON:(MCOAbstractMessage *)msg {
    NSMutableDictionary * email = [self compose: nil];
    NSMutableDictionary * email_headers = [email valueForKey:@"headers"];
    NSMutableDictionary * email_addresses = [email valueForKey:@"addresses"];
    MCOMessageHeader * header = [msg header];
    
    if(header) {
        // Basic data and headers
        if(header.subject) {
            [email setObject:header.subject forKey:@"subject"];
        }
        if(header.date) {
            [email_headers setObject:[header.date description] forKey:@"date"];
        }
        if(header.receivedDate) {
            [email_headers setObject:[header.receivedDate description] forKey:@"received_date"];
        }
        if(header.allExtraHeadersNames) {
            for(NSString * extra in [header allExtraHeadersNames]) {
                NSString * extra_data = [header extraHeaderValueForName:extra];
                if(extra_data) {
                    [email_headers setObject:extra_data forKey:extra];
                }
            }
        }
        // 'From' address
        if(header.from) {
            NSMutableDictionary * from = [email_addresses valueForKey:@"from"];
            NSString * display_name = header.from.displayName;
            NSString * mailbox = header.from.mailbox;
            if(display_name) {
                [from setObject:display_name forKey:@"name"];
            }
            if(mailbox) {
                [from setObject:mailbox forKey:@"mailbox"];
            }
        }
        
        // Remainder of the address types
        NSDictionary * address_sections = @{
                                            @"to": header.to ? header.to : @[],
                                            @"cc": header.cc ? header.cc : @[],
                                            @"bcc": header.bcc ? header.bcc : @[],
                                            @"replyTo": header.replyTo ? header.replyTo : @[]
                                            };
        
        for(NSString * address_section in address_sections.allKeys) {
            NSArray * addresses = [address_sections valueForKey:address_section];
            
            if(addresses) {
                NSMutableArray * my_addresses = [[NSMutableArray alloc] init];
                for(MCOAddress * address in addresses) {
                    NSMutableDictionary * new_address = [[NSMutableDictionary alloc] init];
                    
                    NSString * display_name = address.displayName;
                    NSString * mailbox = address.mailbox;
                    if(display_name) {
                        [new_address setObject:display_name forKey:@"name"];
                    }
                    if(mailbox) {
                        [new_address setObject:mailbox forKey:@"mailbox"];
                    }
                    [my_addresses addObject:new_address];
                }
                [email_addresses setObject:my_addresses forKey:address_section];
            }
        }
    }
    
    return email;
}



- (NSMutableDictionary*)_getEmailStructure {
    return [@{
              @"subject": @"",
              @"body": @""
              } mutableCopy];
}

- (NSMutableDictionary*)_getHeaderStructure {
    return [@{
              } mutableCopy];
}

- (NSMutableDictionary*)_getAddressStructure {
    return [@{
              @"to": [[NSMutableArray alloc] init],
              @"cc": [[NSMutableArray alloc] init],
              @"bcc": [[NSMutableArray alloc] init],
              @"from": [[NSMutableDictionary alloc] init],
              @"replyTo": [[NSMutableDictionary alloc] init]
              } mutableCopy];
}

- (void)_applyHeader:(NSMutableDictionary*)header to:(NSMutableDictionary**)email {
    [*email setObject:header forKey:@"headers"];
}

- (void)_applyAddresses:(NSMutableDictionary*)address to:(NSMutableDictionary**)email {
    [*email setObject:address forKey:@"addresses"];
}


#pragma Public APIs

- (void)getFolders:(id)args {
    ENSURE_ARG_COUNT(args, 2);
    MCOIMAPSession * session = [self _applyCredentials:[args objectAtIndex:0] smpt: NO];
    
    MCOIMAPFetchFoldersOperation * op = [session fetchAllFoldersOperation];
    [op start:^(NSError * error, NSArray *folders) {
        if(error) {
            [[args objectAtIndex:1] call:@[[error description], @[]] thisObject:nil];
        } else {
            NSMutableArray * result = [[NSMutableArray alloc] init];
            for(MCOIMAPFolder * folder in folders) {
                [result addObject: folder.path];
            }
            [[args objectAtIndex:1] call:@[[NSNull null], result] thisObject:nil];
        }
    }];
}

- (void)getFolderInfo:(id)args {
    ENSURE_ARG_COUNT(args, 3);
    MCOIMAPSession * session = [self _applyCredentials:[args objectAtIndex:0] smpt: NO];
    
    MCOIMAPFolderInfoOperation * op = [session folderInfoOperation:[args objectAtIndex:1]];
    [op start:^(NSError * error, MCOIMAPFolderInfo * info) {
        if(error) {
            [[args objectAtIndex:1] call:@[[error description], @[]] thisObject:nil];
        } else {
            NSMutableDictionary * result = [[NSMutableDictionary alloc] init];
            
            [result setObject:[NSNumber numberWithUnsignedLong: [info uidNext]] forKey:@"UIDNEXT"];
            [result setObject:[NSNumber numberWithUnsignedLong: [info uidValidity]] forKey:@"UIDVALIDITY"];
            [result setObject:[NSNumber numberWithUnsignedLong: [info modSequenceValue]] forKey:@"HIGHESTMODSEQ"];
            [result setObject:[NSNumber numberWithInt: [info messageCount]] forKey:@"messages_count"];
            
            [[args objectAtIndex:2] call:@[[NSNull null], result] thisObject:nil];
        }
    }];
}



- (void)getMail:(id)args {
    ENSURE_ARG_COUNT(args, 4);
    MCOIMAPSession * session = [self _applyCredentials:[args objectAtIndex:0] smpt: NO];
    NSString * folder = [args objectAtIndex:1];
    NSArray * range = [args objectAtIndex:2];
    
    MCOIndexSet *uids;
    if(![range isEqual:[NSNull null]]) {
        uids = [MCOIndexSet indexSetWithRange:MCORangeMake([TiUtils intValue:range[0]], [TiUtils intValue:range[1]] - [TiUtils intValue:range[0]])];
    } else {
        uids = [MCOIndexSet indexSetWithRange:MCORangeMake(1, UINT64_MAX)];
    }
    
    MCOIMAPMessagesRequestKind requestKind = (MCOIMAPMessagesRequestKind)
        (MCOIMAPMessagesRequestKindHeaders | MCOIMAPMessagesRequestKindExtraHeaders | MCOIMAPMessagesRequestKindStructure | MCOIMAPMessagesRequestKindHeaderSubject);
    
    MCOIMAPFetchMessagesOperation *fetchOperation = [session fetchMessagesOperationWithFolder:folder requestKind:requestKind uids:uids];
    
	// TODO - extra headers should pass in
    fetchOperation.extraHeaders = @[@"Received-SPF"];
    [fetchOperation start:^(NSError * error, NSArray * fetchedMessages, MCOIndexSet * vanishedMessages) {
        if(error) {
            [[args objectAtIndex:3] call:@[[error description], @[]] thisObject:nil];
        } else {
            NSMutableArray * result = [[NSMutableArray alloc] init];
            for(MCOIMAPMessage * message in fetchedMessages) {
                NSMutableDictionary * email_result = [@{
                                                 @"uid": [NSNumber numberWithInt:message.uid],
                                                 @"sender_name": message.header.sender.displayName ? message.header.sender.displayName : @"",
                                                 @"sender_mailbox": message.header.sender.mailbox ? message.header.sender.mailbox : @"",
                                                 @"subject": message.header.subject ? message.header.subject : @"",
                                                 @"received_time": message.header.receivedDate ? [message.header.receivedDate description] : @"",
                                                 @"has_attachments": message.attachments && [message.attachments count] > 0 ? @true : @false
                                                 } mutableCopy];
                
                for(NSString * hname in message.header.allExtraHeadersNames) {
                    NSString * extra_header = [message.header extraHeaderValueForName:hname];
                    if(extra_header) {
                        [email_result setObject:extra_header forKey:hname];
                    }
                }
                
                [result addObject: email_result];
            }
            [[args objectAtIndex:3] call:@[[NSNull null], result] thisObject:nil];
        }
    }];
}

- (void)getMailInfo:(id)args {
    ENSURE_ARG_COUNT(args, 4);
    MCOIMAPSession * session = [self _applyCredentials:[args objectAtIndex:0] smpt: NO];
    NSString * folder = [args objectAtIndex:1];
    NSInteger * uid = [TiUtils intValue:[args objectAtIndex:2]];
    
    MCOIMAPFetchContentOperation * op = [session fetchMessageOperationWithFolder:folder uid: uid];
    
    [op start:^(NSError * error, NSData * data) {
        if(error) {
            [[args objectAtIndex:2] call:@[[error description], @{}] thisObject:nil];
        } else {
            MCOMessageParser * parser = [MCOMessageParser messageParserWithData:data];
            
            NSMutableDictionary * email = [self _messageToJSON: parser];
            if(parser) {
                [email setObject:[parser htmlBodyRendering] forKey:@"body"];
                
                if(parser.attachments) {
                    NSMutableArray * attachments = [[NSMutableArray alloc] init];
                    for(MCOAttachment * attachment in parser.attachments) {
                       
                        [attachments addObject: @{
                                                  @"file_name": attachment.filename,
                                                  @"mime_type": attachment.mimeType ? attachment.mimeType : @"",
                                                  @"data": attachment.data ? [attachment.data base64Encoding] : @""
                                                  }];
                    }
                    [email setObject: attachments forKey: @"attachments"];
                }
            }
            
            [[args objectAtIndex:3] call:@[[NSNull null], email] thisObject:nil];
        }
    }];
}

- (id)compose: (id)args {
    NSMutableDictionary * email_data = [self _getEmailStructure];
    NSMutableDictionary * email_header = [self _getHeaderStructure];
    NSMutableDictionary * email_address = [self _getAddressStructure];
    
    [self _applyHeader: email_header to:&email_data];
    [self _applyAddresses: email_address to:&email_data];
    
    return email_data;
}


- (id)send: (id)args {
    ENSURE_ARG_COUNT(args, 3);
    NSMutableDictionary * email = [args objectAtIndex:1];
    
    NSMutableDictionary * email_headers = [email valueForKey:@"headers"];
    NSMutableDictionary * email_addresses = [email valueForKey:@"addresses"];
    
    MCOMessageBuilder * builder = [[MCOMessageBuilder alloc] init];
    
    NSString * subject = [email objectForKey: @"subject"];
    NSInteger * receiver_count = 0;
    NSDictionary * from = [email_addresses objectForKey: @"from"];
    receiver_count += [[email_addresses objectForKey: @"to"] count];
    receiver_count += [[email_addresses objectForKey: @"cc"] count];
    receiver_count += [[email_addresses objectForKey: @"bcc"] count];
    if(subject && from && receiver_count > 0) {
        // Subject
        [[builder header] setSubject: subject];
        
        // From
        [[builder header] setFrom: [MCOAddress addressWithDisplayName:[from valueForKey:@"name"] mailbox: [from valueForKey:@"mailbox"]]];
        
        // To
        NSMutableArray * to = [[NSMutableArray alloc] init];
        NSArray * targets = [email_addresses objectForKey: @"to"];
        if(targets) {
            for(int i = 0; i < [targets count]; i++) {
                NSDictionary * target = [targets objectAtIndex: i];
                [to addObject:[MCOAddress addressWithDisplayName:[target valueForKey:@"name"] mailbox:[target valueForKey:@"mailbox"] ]];
            }
            [[builder header] setTo: to];
        }
        
        // cc
        NSMutableArray * cc = [[NSMutableArray alloc] init];
        targets = [email_addresses objectForKey: @"cc"];
        if(targets) {
            for(int i = 0; i < [targets count]; i++) {
                NSDictionary * target = [targets objectAtIndex: i];
                [cc addObject:[MCOAddress addressWithDisplayName:[target valueForKey:@"name"] mailbox:[target valueForKey:@"mailbox"] ]];
            }
            [[builder header] setCc: cc];
        }
        
        // bcc
        NSMutableArray * bcc = [[NSMutableArray alloc] init];
        targets = [email_addresses objectForKey: @"bcc"];
        if(targets) {
            for(int i = 0; i < [targets count]; i++) {
                NSDictionary * target = [targets objectAtIndex: i];
                [bcc addObject:[MCOAddress addressWithDisplayName:[target valueForKey:@"name"] mailbox:[target valueForKey:@"mailbox"] ]];
            }
            [[builder header] setBcc: bcc];
        }
        
        // Copy headers
        for(NSString * key in [email_headers allKeys]) {
            [[builder header] setExtraHeaderValue:[email_headers valueForKey:key] forName:key];
        }
        
        // Email body
        NSString * body = [email valueForKey:@"body"];
        if(body) {
            [builder setHTMLBody: body];
        }
        
        MCOSMTPSession* session = [self _applyCredentials:[args objectAtIndex:0] smpt: YES];
        MCOSMTPSendOperation * op = [session sendOperationWithData:[builder data]];
        [op start:^(NSError *error) {
            if(error) {
                [[args objectAtIndex:2] call:@[[error description], @[]] thisObject:nil];
            } else {
                [[args objectAtIndex:2] call:@[[NSNull null], [self _messageToJSON:builder]] thisObject:nil];
            }
        }];
        
    } else {
        NSLog(@"[ERROR] Cannot send email without a subject, a 'from' section, and some destination");
    }}

@end
